File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Paribu.java
package com.mobnetic.coinguardian.model.market

import com.mobnetic.coinguardian.model.CheckerInfo
import com.mobnetic.coinguardian.model.Market
import com.mobnetic.coinguardian.model.Ticker
import com.mobnetic.coinguardian.model.currency.Currency
import com.mobnetic.coinguardian.model.currency.CurrencyPairsMap
import com.mobnetic.coinguardian.model.currency.VirtualCurrency
import com.mobnetic.coinguardian.util.ParseUtils
import org.json.JSONObject
import java.util.*

class Paribu : Market(NAME, TTS_NAME, CURRENCY_PAIRS) {
    companion object {
        const val NAME = "Paribu"
        const val TTS_NAME = NAME
        const val URL = "https://www.paribu.com/ticker"
        val CURRENCY_PAIRS: CurrencyPairsMap = CurrencyPairsMap()

        init {
            CURRENCY_PAIRS[VirtualCurrency.BTC] = arrayOf(
                    Currency.TRY
            )
        }
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return URL
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val dataJsonObject = jsonObject.getJSONObject("BTC_TL")
        ticker.bid = ParseUtils.getDoubleFromString(dataJsonObject, "highestBid")
        ticker.ask = ParseUtils.getDoubleFromString(dataJsonObject, "lowestAsk")
        ticker.vol = ParseUtils.getDoubleFromString(dataJsonObject, "volume")
        ticker.high = ParseUtils.getDoubleFromString(dataJsonObject, "high24hr")
        ticker.low = ParseUtils.getDoubleFromString(dataJsonObject, "low24hr")
        ticker.last = ParseUtils.getDoubleFromString(dataJsonObject, "last")
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Paribu.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.market.generic.SimpleMarket
import com.aneonex.bitcoinchecker.datamodule.util.forEachName
import org.json.JSONObject

class Paribu : SimpleMarket(
    "Paribu",
    URL,
    URL, // TODO: There is no API for a separate market yet
) {
    companion object {
        private const val URL = "https://v3.paribu.com/app/ticker"
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return URL
    }

    override fun parseCurrencyPairsFromJsonObject(
        requestId: Int,
        jsonObject: JSONObject,
        pairs: MutableList<CurrencyPairInfo>
    ) {
        jsonObject
            .getJSONObject("data")
            .forEachName { market, _ ->
                val assets = market.uppercase().split('-')
                if(assets.size == 2){
                    pairs.add(CurrencyPairInfo(
                        assets[0],
                        assets[1],
                        null
                    ))
                }
            }
    }

    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val market = with(checkerInfo){"$currencyBase-$currencyCounter"}
        val dataJsonObject = jsonObject
            .getJSONObject("data")
            .let { data ->
                data.optJSONObject(market) ?: data.getJSONObject(market.lowercase())
            }

        dataJsonObject.also {
            ticker.bid = it.getDouble("bid")
            ticker.ask = it.getDouble("ask")

            ticker.high = it.getDouble("high")
            ticker.low = it.getDouble("low")

            ticker.vol = it.getDouble("volume")
            ticker.last = it.getDouble("close")
        }
    }
}